### ISSUE TITLE
Meta tag index corruption can happen when updating meta records via swap

### ISSUE THREAD

**Issue by replay** [grafana/metrictank#1661]
This has been reported to lead to a panic in certain queries involving that corrupted index.

The panic happened here after calling swap with an empty set of meta records (deleting all records): https://github.com/grafana/metrictank/blob/master/idx/memory/tag_query_id_filter.go#L158

I suspect that some meta tags must have not been cleaned up from the meta tag index correctly during the delete.

**Comment by replay** [comment]
I have a suspicion what might have happened. 

For this panic at https://github.com/grafana/metrictank/blob/9b7ee522a34c8ad1f5f65d5aff1973ecac337d72/idx/memory/tag_query_id_filter.go#L158
to get triggered, a meta record with `0` expressions must have been passed into the function `metaRecordFilterBySetOfValidValues()`. A meta record with `0` expressions should never be accepted by the endpoints which allow the user to update the meta records, a request posting such a meta record would get rejected. So I suspect that when the set of meta records to pass into this function got prepared this case got hit https://github.com/grafana/metrictank/blob/9b7ee522a34c8ad1f5f65d5aff1973ecac337d72/idx/memory/tag_query_id_filter.go#L97 which results in a `continue` of that loop, but the slice of meta records (`records`) has already been pre-allocated at this point, resulting in one empty meta record which would have no expressions. 

I don't know yet how this could have happened the index ended up in this state, so for now I just added some additional protections to prevent a panic and I'm still looking into how exactly this happened: https://github.com/grafana/metrictank/pull/1662

**Comment by replay** [comment]
I found one way how the index could have ended up in a state where this case would get triggered https://github.com/grafana/metrictank/blob/9b7ee522a34c8ad1f5f65d5aff1973ecac337d72/idx/memory/tag_query_id_filter.go#L97

When a swap operation gets executed by the memory index, these are the steps which get executed:
* provided meta records get filtered to remove all the ones which are already present
* remaining ones get inserted (if another existing one has the same expressions it gets deleted/replaced)
* all the ones which have not been inserted/replaced at the previous step get pruned out
  * the pruning iterates over record by record
  * for each record it first removes it from the `metaTagRecords` data structures first
  * then it deletes the associations to that record from the `metaTagIndex` (with a separate lock from the previous step)

So to fix this we could just reverse the order of the last two steps, I prefer to not hold the lock across both steps to not cause unnecessary lock congestion. But for now the PR #1662 should be sufficient to prevent a similar panic, and this described issue is only a temporary inconsistency until the swap operation completed executing.

**Comment by shanson7** [comment]
In our case, we got this panic consistently for hours after the swap completed.

**Comment by shanson7** [comment]
Here is a full stack of the panic: https://gist.github.com/shanson7/ced051cf78d164b2d0ab6cfadceffad3

Somehow I missed it previously, but there is a clear "Index is corrupted" message at the top.

**Comment by replay** [comment]
Thx @shanson7. At least this dump confirms that the index was corrupted in that way which i suspected. I'll further investigate into how it ended up in this state

**Comment by replay** [comment]
Another possibility how the index could get into this state would be if during the pruning an invalid meta record gets detected, because the pruning first prunes the `metaTagRecords` structure:

https://github.com/grafana/metrictank/blob/master/idx/memory/memory.go#L733

And then it prunes the `metaTagIndex` structure:

https://github.com/grafana/metrictank/blob/master/idx/memory/memory.go#L751

But in-between it instantiates a query from the record, which has the potential to fail if the record is invalid:

https://github.com/grafana/metrictank/blob/master/idx/memory/memory.go#L741

If that error on line `:741` occurs then it would `continue` in the loop and omit the pruning from the `metaTagIndex`. I can't see how such an invalid record could get inserted into the index, because before they get inserted they get validated in the same way, the only way would be if it somehow gets modified in the memory after inserting it.

@shanson7 do you still have the logs of that instance which panicked? before the panics started, do you have this log line? https://github.com/grafana/metrictank/blob/master/idx/memory/memory.go#L742

Either way it should be easy to make sure that the `metaTagIndex` gets pruned anyway, even if the instantiation of that query fails.

**Comment by agao48** [comment]
We still have the logs from the panic - didn't see that log line anywhere.

**Comment by replay** [comment]
I think I might have found the issue. Obviously I can't prove that this is how your index got corrupted, but it is one possibility how the index would end up in a state which would have resulted in the panic that you saw, and it wouldn't be transient, it would be permanent.

When a swap operation happens and a meta record already exists in the index, but with different meta tags (remember, the ID of a record is its set of expressions), then the swap procedure first removes the references to the old record from the `metaTagIndex`, those references are keyed by the meta tags associated with the record. Then it updates the meta record with the new meta tags in the `metaTagRecords`. Then it creates new references to it in the `metaTagIndex` (keyed by the new meta tags). 

This first deletion of the references from the `metaTagIndex` happens here: https://github.com/grafana/metrictank/blob/master/idx/memory/memory.go#L689

But the issue is that instead of iterating over the meta tags of the old record, this iterates over the records of the new updated record. So if the old record had meta tags which don't exist in the updated one or have a different value in the updated one, then the references to the old record do not get cleaned up properly. In most cases this wouldn't become obvious right away, because most likely the updated meta record is going to get the same ID in the `metaTagRecords` when it gets replaced there. But once you wipe the index by calling a swap with an empty set of records, those references from the `metaTagIndex` are not going to get cleaned up, because they shouldn't be there.

I was able to reproduce it like this:

* check out version `bc00db9a` (before the patches i recently added to protect against panic)
* insert one metric: `some.id.of.a.metric.1;afewmoretags=forgoodmeasure`
* call swap with `{"records":[{"expressions":["afewmoretags=forgoodmeasure"],"metaTags":["a=a"]}]}'` to add a meta tag `a=a`
* call swap with `{"records":[{"expressions":["afewmoretags=forgoodmeasure"],"metaTags":["a=b"]}]}'` to update the meta record. This should remove the reference keyed `a=a` from `metaTagIndex`, but doesn't due to the above described bug
* call swap with `{}` to remove all entries from the `metaTagRecords`, now there is a reference in the `metaTagIndex` which refers to a non-existent record id
* query `target=seriesByTag("name=some.id.of.a.metric.1","a=a")'` to hit that reference that wasn't cleaned up

same panic:

```
2020-02-06 15:39:47.442 [ERROR] TagQueryContext: Tried to lookup a meta tag record id that does not exist, index is corrupted
[Macaron] PANIC: runtime error: index out of range [0] with length 0
/usr/local/go/src/runtime/panic.go:75 (0x42daf2)
/home/replay/go/src/github.com/grafana/metrictank/idx/memory/tag_query_id_filter.go:158 (0xce0685)
/home/replay/go/src/github.com/grafana/metrictank/idx/memory/tag_query_id_filter.go:128 (0xce00a9)
/home/replay/go/src/github.com/grafana/metrictank/idx/memory/tag_query.go:144 (0xcdf192)
/home/replay/go/src/github.com/grafana/metrictank/idx/memory/tag_query.go:210 (0xcdf801)
/home/replay/go/src/github.com/grafana/metrictank/idx/memory/tag_query.go:187 (0xcdf6d0)
/home/replay/go/src/github.com/grafana/metrictank/idx/memory/memory.go:1528 (0xcce4cb)
/home/replay/go/src/github.com/grafana/metrictank/idx/memory/memory.go:1075 (0xcc93b1)
/home/replay/go/src/github.com/grafana/metrictank/api/cluster.go:329 (0xbbeca7)
/usr/local/go/src/reflect/value.go:460 (0x4b1935)
....
```

**Comment by replay** [comment]
I think for now it's safe to assume that #1667 fixes this issue. Otherwise, if this happens again because the corruption happened to to another bug than the one fixed in #1667, then we can reopen this.
